### DISCUSSION TITLE
Changing Work Mode from HA

### DISCUSSION THREAD

**Discussion by wepster** [davidrapan/ha-solarman#112] [General]
Hi

I'm using this integration to monitor my deye 8k 3p inverter.
I would like to change the main Work Mode from "zero export" to "selling first" from automation (to sell some electricity from battery when price is highest).
I have 163 entities using deye_sg04lp3.yaml inverter definition but I'm unable to do it.
Do you know of any way to do this (like adding another switch with correct registers)?
As far as i can tell, changing to "selling first" is mandatory to be able to sell power from battery, and zero export with selling rusplus solar enabled is what i want to use normally. 
And i dont see any way to change iwork modet from HA.
Please advise :)

**Comment by davidrapan** [comment]
Hey @wepster, do you have any idea which register it is?

**Comment by wepster** [comment]
No, that would solve my problem, currently im waiting for documentation from deye support.
One of device definitions (deye hybrid i belive) includes registry number for this but only to read, however in mine inverter there is no thing under this address. 
I was unable to find ant other address to check.

**Reply by davidrapan** [reply]
What device model do you have?

**Comment by wepster** [comment]
Sun-8k-sg04lp3-eu

**Comment by PaulOckenden** [comment]
See if 4608 (corresponds to 0x1200) changes when you set different work modes on your inverter. That's the register that's used on my Sofar, so maybe your Sun uses the same place?

Probably not, but it's worth a try.

**Comment by wepster** [comment]
Unfortunately 0x1200 makes configuration incorrect and integration doesnt initialize.
Support mailed me some documentation (attachment) but i dont see anything helpful in it.
[deye.zip](https://github.com/user-attachments/files/16795642/deye.zip)

**Comment by PaulOckenden** [comment]
Looks like the registers 141 and upwards might be useful. Can you map how these relate to how you'd make the changes you need via the user interface? It's difficult to comment further without knowing how this particular inverter works - they all have slightly different controls.

**Reply by davidrapan** [reply]
141 Is already controllable using integration natively as Energy Pattern with bit0-1 options Load First and Battery First.

_When comes to the other bits I was not yet able to determine their functionality by changing the setting on the device itself._

**Comment by wepster** [comment]
i will try to "Scan" more registers and observe them when i change teh setting from menu.

**Comment by wepster** [comment]
I am unable to scan all registers, cant find working tool for it (like modbustester).
Deye support send me a document i attached earlier, and i cant find appropriate register address in it, and Deye support tells me to ask company who installed the inverter, but they don't know it either and tell me to contat Deye.
Unfortunatelly i don't know what to do with this.

**Reply by davidrapan** [reply]
Yeah I also found dead end while trying to find the proper place for it :-/

**Comment by wepster** [comment]
DONE :)

      - name: WorkMode
        update_interval: 10
        platform: "select"
        state_class: ""
        uom: ""
        scale: 1
        rule: 1
        registers: [0x008E]
        options: ["Selling First", "Zero Export To Load", "Zero Export To CT"]
        lookup:
          - key: 0x0000
            value: "Selling First"
          - key: 0x0001
            value: "Zero Export To Load"
          - key: 0x0002
            value: "Zero Export To CT"

I'm sure i tried this register before i started this thread, but maybe i did something wrong then.
Or, DEYE patched it with newest software update.
Since the update (couple days now) i can discharge battery down do 20%, before this update lowest was 35%. 
Glad to have done that, now i can autmate energy selling.
Thank you guys fo your effort.

**Reply by davidrapan** [reply]
Great, I'll add it. 😉

**Reply by wepster** [reply]
with today's update installation my custom_component.yaml got deleted
not cool, please dont do that :)

**Reply by davidrapan** [reply]
Point number 1. in https://github.com/davidrapan/ha-solarman/wiki#-changes 😉

**Reply by wepster** [reply]
After todays update from 24.11
![image](https://github.com/user-attachments/assets/d3ac43c4-d912-48c2-8599-b4eea1063847)

**Reply by davidrapan** [reply]
I always have some files in custom folder to check exactly bugs like and can say that integration does not touch them. 😉

But w/ [v24.12.08_7](https://github.com/davidrapan/ha-solarman/releases/tag/v24.12.08_7) there is transition from Github Source code releases to zip releases for HACS so that's the only thing that comes to mind what could be the culprit.

Edit: Can you do partial restore from backup and try to update to version before that?

**Reply by wepster** [reply]
unfortunatelly update did't offer otpion (no checkbox) to create backup, so i dont have one.
I knew i would not be able to be much help, thats why i posted it here instead of creating issue.
This integration is a life saver for me, thank you for your effort.

**Reply by davidrapan** [reply]
Mh, I just did restore and update back and forth twice and custom folder is still intact... I am really confused about what happened in your case...

**Reply by jpsl1979** [reply]
> HECHO :)
> 
> ```
>   - name: WorkMode
>     update_interval: 10
>     platform: "select"
>     state_class: ""
>     uom: ""
>     scale: 1
>     rule: 1
>     registers: [0x008E]
>     options: ["Selling First", "Zero Export To Load", "Zero Export To CT"]
>     lookup:
>       - key: 0x0000
>         value: "Selling First"
>       - key: 0x0001
>         value: "Zero Export To Load"
>       - key: 0x0002
>         value: "Zero Export To CT"
> ```
> 
> Estoy seguro de que probé este registro antes de empezar este hilo, pero quizás hice algo mal entonces. O DEYE lo parcheó con la última actualización de software. Desde la actualización (hace un par de días), puedo descargar la batería hasta un 20 %, antes de esta actualización, el mínimo era del 35 %. Me alegra haberlo hecho; ahora puedo automatizar la venta de energía. Gracias por su esfuerzo.

HOla! yo estoy intentando descargar batería a red también con mi turbo energy DEYE, pero no hay manera, tengo en cuenta todas las opciones según modbus v118 y nada, que versión de firmware tienes en el inversor yo, 
Turbo Energy / Deye LP1 monofásico híbrido (familia SG0*LP1), con versiones HMI 8347, MAIN1 3386 y MAIN2 513.
En mi caso creo que el problema está en el firmware, por eso te preguntaba, un saludo y gracias!!
